core/modules/loader/MWorkerList.cc

```diff
@@ -0,0 +1,423 @@
+        }
+        std::weak_ptr<SetNeighborOneShot> oneShotData;
+    };
+    auto ptr = std::dynamic_pointer_cast<SetNeighborOneShot>(getDoListItemPtr());
+    return std::make_shared<SetNeighborCmd>(ptr);
+}
+
+
+util::CommandTracked::Ptr MWorkerList::createCommand() {
+    return createCommandMaster(_central);
+}
+
+
+util::CommandTracked::Ptr MWorkerList::createCommandMaster(CentralMaster* centralM) {
+    // The master probably doesn't need to make any checks on the list.
+    return nullptr;
+}
+
+
+// Returns pointer to new item when new worker added, otherwise nullptr.
+MWorkerListItem::Ptr MWorkerList::addWorker(std::string const& ip, int udpPort, int tcpPort) {
+    NetworkAddress udpAddress(ip, udpPort);
+    NetworkAddress tcpAddress(ip, tcpPort);
+
+
+    // If it is already in the map, do not change its name.
+    std::lock_guard<std::mutex> lock(_mapMtx);
+    auto iter = _ipMap.find(udpAddress);
+    if (iter != _ipMap.end()) {
+        LOGS(_log, LOG_LVL_WARN, "addWorker, Could not add worker as worker already exists. " <<
+                ip << ":" << udpPort);
+        return nullptr;
+    }
+    // Get an id and make new worker item
+    auto workerListItem = MWorkerListItem::create(_sequenceName++, udpAddress, tcpAddress, _central);
+    _ipMap.insert(std::make_pair(udpAddress, workerListItem));
+    _wIdMap.insert(std::make_pair(workerListItem->getId(), workerListItem));
+    LOGS(_log, LOG_LVL_INFO, "Added worker " << *workerListItem);
+    _flagListChange();
+
+    return workerListItem;
+}
+
+bool MWorkerList::sendListTo(uint64_t msgId, std::string const& ip, short port,
+                            std::string const& ourHostName, short ourPort) {
+    NetworkAddress address(ip, port);
+    StringElement workerList;
+    {
+        std::lock_guard<std::mutex> lock(_mapMtx);
+        if (_wListChanged || _stateListData == nullptr) {
+            _wListChanged = false;
+            /// At this time, all workers should easily fit in a single message.
+            /// TODO send multiple messages (if needed) with each having the address and range of 100 workers.
+            ///      This version is useful for testing. _stateListData becomes a vector.
+            proto::LdrMastWorkerList protoList;
+            protoList.set_workercount(_wIdMap.size());
+            for (auto const& item : _wIdMap ) {
+                proto::WorkerListItem* protoItem = protoList.add_worker();
+                MWorkerListItem::Ptr const& wListItem = item.second;
+                protoItem->set_wid(wListItem->getId());
+            }
+            protoList.SerializeToString(&(workerList.element));
+            LoaderMsg workerListMsg(LoaderMsg::MAST_WORKER_LIST, msgId, ourHostName, ourPort);
+            _stateListData = std::make_shared<BufferUdp>();
+            workerListMsg.appendToData(*_stateListData);
+            workerList.appendToData(*_stateListData);
+        }
+    }
+
+    _central->sendBufferTo(ip, port, *_stateListData);
```

Comment:
I think this is unsafe; this could:
1. release the lock_guard at line 180
2. pass the reference to the managed object into sendBufferTo on this line.
3. another thread enters this function and assigns a new object to _stateListData, which causes the previously referred-to object to be destroyed.
4. somewhere down the line inside of `workerListMsg.appendToData` there is a segfault.